source/extensions/filters/http/decompressor/decompressor_filter.cc

```diff
@@ -0,0 +1,213 @@
+  //      the upstream that this hop is able to decompress responses via the Accept-Encoding header.
+  if (config_->responseDirectionConfig().decompressionEnabled()) {
+    if (config_->requestDirectionConfig().advertiseAcceptEncoding()) {
+      headers.appendAcceptEncoding(config_->contentEncoding(), ",");
+      ENVOY_STREAM_LOG(
+          debug, "DecompressorFilter::decodeHeaders advertise Accept-Encoding with value '{}'",
+          *decoder_callbacks_, headers.AcceptEncoding()->value().getStringView());
+    }
+  }
+
+  //   2. If request decompression is enabled, then decompress the request.
+  return maybeInitDecompress(config_->requestDirectionConfig(), request_decompressor_,
+                             *decoder_callbacks_, headers);
+};
+
+Http::FilterDataStatus DecompressorFilter::decodeData(Buffer::Instance& data, bool end_stream) {
+  return maybeDecompress(config_->requestDirectionConfig(), request_decompressor_.get(),
+                         *decoder_callbacks_, data, end_stream);
+}
+
+Http::FilterHeadersStatus DecompressorFilter::encodeHeaders(Http::ResponseHeaderMap& headers,
+                                                            bool end_stream) {
+  // Headers only response, continue.
+  if (end_stream) {
+    return Http::FilterHeadersStatus::Continue;
+  }
+  ENVOY_STREAM_LOG(debug, "DecompressorFilter::encodeHeaders: {}", *encoder_callbacks_, headers);
+
+  return maybeInitDecompress(config_->responseDirectionConfig(), response_decompressor_,
+                             *encoder_callbacks_, headers);
+}
+
+Http::FilterDataStatus DecompressorFilter::encodeData(Buffer::Instance& data, bool end_stream) {
+  return maybeDecompress(config_->requestDirectionConfig(), response_decompressor_.get(),
+                         *decoder_callbacks_, data, end_stream);
+}
+
+Http::FilterHeadersStatus
+DecompressorFilter::maybeInitDecompress(DecompressorFilterConfig::DirectionConfig& direction_config,
+                                        Compression::Decompressor::DecompressorPtr& decompressor,
+                                        Http::StreamFilterCallbacks& callbacks,
+                                        Http::RequestOrResponseHeaderMap& headers) {
+  if (direction_config.decompressionEnabled() && !hasCacheControlNoTransform(headers) &&
+      contentEncodingMatches(headers)) {
+    direction_config.stats().decompressed_.inc();
+    decompressor = config_->makeDecompressor();
+    // remove the compression scheme's content encoding because the filter is going to decompress.
+    removeContentEncoding(headers);
+
+    // Per the filter documentation, only buffer if there is a Content-Length header present, and
+    // the filter is configured to buffer.
+    if (headers.removeContentLength() > 0 && direction_config.maxBufferedBytes()) {
+      direction_config.setBuffering(true);
+      callbacks.setBufferLimit(direction_config.maxBufferedBytes().value());
```

Comment:
@alyssawilk I have this logic that acts on streams regardless of request/response. I need to interact with the stream callbacks and I was wondering why we don't have a virtual `setBufferLimit` function instead (or in addition to) the `setDecoder/EncoderBufferLimit` in the derived classes?

Reply:
If you look at what's in StreamFilterCallbacks it affects everything about the stream.  There's one stream ID.  If you reset the stream it is reset.  if you SetBufferLimits on the stream and it only sets the request or response buffer limits IMO it's confusing.    In general it's better to have the clarity on what limits you're setting so folks can tell if they're affecting the request or response path even if it involves some down casts or absl::variants, but that's just my opinion.

Reply:
Thanks for the explanation! Makes sense, all functionality in the base class affects the stream as a whole, not individual directions. I'll rework this.